webapp/packages/core-blocks/src/useSuspense.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { action, IReactionDisposer, observable, reaction } from 'mobx';
+
+import { useObservableRef } from './useObservableRef';
+
+interface IObservedValueMetadata<TArgs, TValue> {
+  reaction: IReactionDisposer | null;
+  promise: Promise<TValue> | null;
+  error: Error | null;
+  value: TValue | symbol;
+  version: number;
+  run(args: TArgs): void;
+}
+
+interface ISuspense {
+  observedValue<TArgs, TValue>(key: string, args: () => TArgs, loader: (args: TArgs) => Promise<TValue>): () => TValue;
+}
+
+interface ISuspenseState extends ISuspense {
+  observedValueMetadata: Map<string, IObservedValueMetadata<any, any>>;
+}
+
+const VALUE_NOT_SET = Symbol('value not set');
+
+/**
+ * Experimental, use to pass suspended value getter to the child components
+ *
+ * (!!!) Don't access the suspended value in the same component where useSuspense is declared
+ * @returns
+ */
+export function useSuspense(): ISuspense {
+  const state = useObservableRef<ISuspenseState>(
+    () => ({
+      observedValueMetadata: new Map(),
+
+      observedValue<TArgs, TValue>(key: string, args: () => TArgs, loader: (args: TArgs) => Promise<TValue>): () => TValue {
```

Comment:
we can consider the `key` argument to be a symbol, not a string. to avoid collisions when the same key can has 2 different values depending on context where it was called

Reply:
there is no difference because the main problem is that the same component can be mounted more than 1 time and will use the same key (no matter the type)